File: Solutions/1-two-sum/two-sum.ex
defmodule Solution do
  @spec two_sum(nums :: [integer], target :: integer) :: [integer]
  def two_sum(nums, target) do
    two_sum(nums, target, %{}, 0)
  end

  defp two_sum([], _, _, _), do: []  # If no pair is found, return empty list

  defp two_sum([head | tail], target, map, index) do
    complement = target - head  # Calculate complement
    case Map.get(map, complement) do
      nil -> # If complement is not found, add number and index to map
        two_sum(tail, target, Map.put(map, head, index), index + 1)
      found_index -> # If complement found, return the pair of indices
        [found_index, index]
    end
  end
end

File: Solutions/1-two-sum/two-sum.erl
-spec two_sum(Nums :: [integer()], Target :: integer()) -> [integer()].
two_sum(Nums, Target) ->
  % Create an empty hash table
  two_sum(Nums, Target, #{}, 0).

two_sum([], _, _, _) -> 
  []; % If no pair is found, return empty list

two_sum([Head | Tail], Target, Map, Index) ->
  Complement = Target - Head, % Calculate complement
  case maps:get(Complement, Map, undefined) of
    undefined -> % If complement not found, add number and index to hash table
      two_sum(Tail, Target, maps:put(Head, Index, Map), Index + 1);
    FoundIndex -> % If complement found, return the pair of indices
      [FoundIndex, Index]
  end.
